### ISSUE TITLE
submitMultipleOrders doesn't work right

### ISSUE THREAD

**Issue by raymondjhs** [tiagosiebler/binance#162]
Hello,
The submitMultipleOrders doesn't work right. Can you help to find how to setup the right params?
Below the is the test code.

`const client = new USDMClient({
    api_key: API_KEY,
    api_secret: API_SECRET,
});

client.submitMultipleOrders([{ symbol: 'BTCUSDT', side: 'BUY', type: 'LIMIT', quantity: 0.001, price: 45000, timeInforce: 'GTC' }, { symbol: 'ETHUSDT', side: 'SELL', type: 'LIMIT', quantity: 0.01, price: 6000, timeInforce: 'GTC' }])
    .then(result => {
        console.log("submitNewOrder result: ", result);
    })
    .catch(err => {
        console.error("submitNewOrder error: ", err);
    });`

**Comment by A-ndy-git** [comment]
See the required params in the official Binance documentation [Binance API Docs](https://binance-docs.github.io/apidocs/futures/en/#place-multiple-orders-trade).

Can you please provide the error message returned from the Binance client?

**Comment by raymondjhs** [comment]
> See the required params in the official Binance documentation [Binance API Docs](https://binance-docs.github.io/apidocs/futures/en/#place-multiple-orders-trade).
> 
> Can you please provide the error message returned from the Binance client?

There are two type of error message.

1.some times, there is the error when I submit single order or batch orders, but it does not always happened. I synced my computer's time, this problem seemed solved:
  code: -1021,
  message: "Timestamp for this request was 1000ms ahead of the server's time.",

2.submitMultipleOrders error (this error don't happend for sumbitNewOrder):  
{
  code: -1022,
  message: 'Signature for this request is not valid.',
  body: { code: -1022, msg: 'Signature for this request is not valid.' },
  headers: {
    'content-type': 'application/json',
    'content-length': '63',
    connection: 'close',
    date: 'Fri, 19 Nov 2021 04:04:17 GMT',
    server: 'Tengine',
    'x-mbx-used-weight-1m': '5',
    'x-response-time': '1ms',
    'access-control-allow-origin': '*',
    'access-control-allow-methods': 'GET, POST, PUT, DELETE, OPTIONS',
    'x-cache': 'Error from cloudfront',
    via: '1.1 bcaa73dd9eb58937e5bb422fc78109e1.cloudfront.net (CloudFront)',
    'x-amz-cf-pop': 'NRT57-C2',
    'x-amz-cf-id': 'Mf__4GjkjLp8x56PUhkKpXFTYTxYOJ1eFLdPGBMyN9yjnrvN8jtsPg=='
  },
  requestUrl: 'https://fapi.binance.com/fapi/v1/batchOrders',
  requestBody: undefined,
  requestOptions: {
    recvWindow: 5000,
    syncIntervalMs: 3600000,
    strictParamValidation: false,
    api_key: undefined,
    api_secret: undefined
  }
}

**Comment by tiagosiebler** [comment]
> > See the required params in the official Binance documentation [Binance API Docs](https://binance-docs.github.io/apidocs/futures/en/#place-multiple-orders-trade).
> > Can you please provide the error message returned from the Binance client?
> 
> There are two type of error message.
> 
> 1.some times, there is the error when I submit single order or batch orders, but it does not always happened. I synced my computer's time, this problem seemed solved: code: -1021, message: "Timestamp for this request was 1000ms ahead of the server's time.",
> 
> 2.submitMultipleOrders error (this error don't happend for sumbitNewOrder): { code: -1022, message: 'Signature for this request is not valid.', body: { code: -1022, msg: 'Signature for this request is not valid.' }, headers: { 'content-type': 'application/json', 'content-length': '63', connection: 'close', date: 'Fri, 19 Nov 2021 04:04:17 GMT', server: 'Tengine', 'x-mbx-used-weight-1m': '5', 'x-response-time': '1ms', 'access-control-allow-origin': '*', 'access-control-allow-methods': 'GET, POST, PUT, DELETE, OPTIONS', 'x-cache': 'Error from cloudfront', via: '1.1 bcaa73dd9eb58937e5bb422fc78109e1.cloudfront.net (CloudFront)', 'x-amz-cf-pop': 'NRT57-C2', 'x-amz-cf-id': 'Mf__4GjkjLp8x56PUhkKpXFTYTxYOJ1eFLdPGBMyN9yjnrvN8jtsPg==' }, requestUrl: 'https://fapi.binance.com/fapi/v1/batchOrders', requestBody: undefined, requestOptions: { recvWindow: 5000, syncIntervalMs: 3600000, strictParamValidation: false, api_key: undefined, api_secret: undefined } }

For the first error, your approach was correct in making sure your computer's time is in sync. The fact the timestamp was ahead was also a side effect of the "time sync" feature, which you can read about here, and disable if you like:
https://github.com/tiagosiebler/binance/issues/153#issuecomment-947018414

I'm currently evaluating having it disabled by default as I'm not sure it's solving more problems than it's causing (as one slow request can impact the timestamp for future requests, causing the error that you saw with the timestamp).

For the second error, I'll look into it. Unusual to see a signature not valid error, definitely not expected. I'll try it on my end to see if it reproduces the error.

**Comment by tiagosiebler** [comment]
I'm also seeing the error "Signature for this request is not valid." when trying to submit multiple orders at once. The same request works in a call to submitNewOrder() once per order. Will need to investigate - if you find the cause before me, please let me know. PRs also welcome.

**Comment by tiagosiebler** [comment]
It's strange, the syntax isn't quite clear to me. This is how I'm testing, in case you wish to join the effort:
```
const client = new USDMClient({
  api_key: key,
  api_secret: secret,
  beautifyResponses: true,
  disableTimeSync: true,
});

(async () => {
  try {

    const order1 = {
      symbol: 'BTCUSDT',
      price: 55499,
      quantity: 0.001,
      side: 'BUY',
      type: 'LIMIT',
      timeInForce: 'GTC',
    };

    const requestParams = {
      // batchOrders: [ orderJsonObject1 ],
      // batchOrders: JSON.stringify([orderJsonObject1]),
      batchOrders: encodeURIComponent(JSON.stringify([order1])),
    };

    console.log('reqParamsBeforeSign: ', requestParams);
    const res = await client._call('POST', 'fapi/v1/batchOrders', requestParams, true);
    
    // const res = await client.submitNewOrder(order);
    // const res = await client.submitMultipleOrders([order]);
    console.log('res: ', JSON.stringify(res, null, 2));

  } catch (e) {
    console.error('request failed: ', e);
  }
})();
```

Comments in the binance-api telegram group suggested url-encoding the json that goes into the batchOrders param, which is the most recent attempt. Only if I set the batchorders parameter to false or remove it does the api stop complaining about signature. Seems odd to have to change the signature process for one unique request type though, so I'm not sure that's it...

**Comment by tiagosiebler** [comment]
That's great! Thank you for this. Will look to automate this in the wrapper implementation, so you can just pass an array of order objects and don't have to worry about the list syntax used by this endpoint.

**Comment by raymondjhs** [comment]
> Try use below code. This solved the bug.
> 
> ```
> let batchOrders = [];
> for(let i=1; i< 3; i++) {
>     batchOrders.push(JSON.stringify({
>         symbol: 'BTCUSDT',
>         side: 'BUY',
>         positionSide: 'BOTH',
>         type: 'LIMIT',
>         timeInForce: 'GTC',
>         quantity: '0.001',
>     }))
> }
> 
> binanceUSDMClient.postPrivate('/fapi/v1/batchOrders', {
>     batchOrders : '['+batchOrders.join(',')+']'
> }).then(result => {
>     console.log(result);
> }).catch(err => {
>     console.error(err);
> });
> ```

Good job. Thank you very much. I think the doucuments of binance API about this point isn't clear. There is also similar problem when cancel batch orders.

**Comment by tiagosiebler** [comment]
Fixed in v2.0.32. Thanks for everyone that helped here.
